Infiniteprogressbar

paulhovey opened the pull request:
Implements #108 
I made a small goroutine to constantly update the progress bar, and a constant rate of refresh defined at 50ms.
The regular progress bar should function exactly the same.  New functions of interest to users are `NewInfiniteProgressBar()` and `StopInfiniteProgress()` .  They seem kinda wordy to me, but I thought abbreviating reduced readability and broke styling from the rest of the project.
I added a basic test to make sure the new variables distinguishing an infinite progress bar from a normal one are set correctly on both.  Also added an infinite bar to the demo underneath the regular progress bar, and it stops looping when the regular one hits 100%.

andydotxyz commented:
Thanks for this, the code is clear and neat and the use of a channel looks good.

Unfortunately I think we need to break a little more from the traditional progress bar - an infinite (or indeterminate?) bar should not look like it could complete. What we could create here is frustration as it effectively goes 100% -> 0%. Maybe the "progress" chunk could expand to, for example, 10% and then move along the bar? Check out this document https://uxplanet.org/progress-indicators-in-mobile-ux-design-a141e22f3ea0

Additionally if we separate them a little more - such that they are not just the same type:
* The infinite no longer has Min/Max/Value (which don't seem necessary)
* The StopInfiniteProgress() could be just Stop()
* It removes the logging if nil channel as we can be sure this never happens

Let me know what you think about this. It's possible that they share a renderer to avoid duplicate code - but I think their public APIs may need to split a little.
With an "Infinite" field users could make this into an infinite bar, or back, at any time - we should decide if that should be supported or not.

steveoc64 commented on widget/progressbar.go:
```diff
@@ -114,15 +122,55 @@ func (p *ProgressBar) SetValue(v float64) {
+		return
+	}
+
+	tickChan := time.NewTicker(infiniteRefreshRate).C
```
ticker := time.NewTicker(infiniteRefreshRate)
defer ticker.Stop()

then use `case <-ticker.C` below

Otherwise - Im pretty sure ... each call to infiniteProgressLoop() will leave a goroutine running that continues to emit ticks.   eeeks !

steveoc64 commented on widget/progressbar.go:
```diff
@@ -74,10 +79,13 @@ func (p *progressRenderer) Objects() []fyne.CanvasObject {
-	Min, Max, Value float64
+	Min, Max, Value      float64
+	Infinite             bool
+	stopInfiniteLoopChan chan bool
```
very slightly more efficient - can use the (ugly) empty struct channel as a signalling mechanism, where you dont care what the actual value is.

In this case, the consumer of the channel doesnt care if the value is true of false, so its a good candidate for this hack.

```
stopInfiniteLoopChan chan struct{}

then on 128 below

p.stopInfiniteLoopChan <- struct{}{}
```

paulhovey replied:
I ended up restructuring a little, but cool to know about the empty signal :)

paulhovey commented:
Update:  I got my laptop back 😄 

I put the infinite progress bar into it's own file, it seemed cleaner than having `if bar.Infinite {}` in a bunch of functions.  Tests can be separated also.

Since the infinite progress bar was separated from the regular progress bar, I removed the channel that was there before, and instead made a ticker an internal variable that can be called with `Stop()`.  Seemed cleaner than a channel for this simple loop.

I feel as though there's a bunch of repetitive logic in `updateBar()`, but wasn't sure how to restructure to make it more efficient within the framework.

steveoc64 commented:
Nice work Paul :) 

Ticker update looks good.  Its a common cause of resource leaks that one.

paulhovey commented:
> Nice work Paul :)
> 
> Ticker update looks good. Its a common cause of resource leaks that one.

Thank you, that reminded me!  When testing on the demo app, I found that if I navigate to a different tab, then `Hide()` was called on the progress bar widget.  However, if I exited by clicking the 'x', then the background ticker kept going because `Hide()` was never called.  Is there a common hook that's passed into a widget when a window is closed?

steveoc64 commented:
> > Nice work Paul :)
> > Ticker update looks good. Its a common cause of resource leaks that one.
> 
> Thank you, that reminded me! When testing on the demo app, I found that if I navigate to a different tab, then `Hide()` was called on the progress bar widget. However, if I exited by clicking the 'x', then the background ticker kept going because `Hide()` was never called. Is there a common hook that's passed into a widget when a window is closed?

Good Q.  It would be reasonable to assume that an OnClose() call would exist.  Raise the Q in the slack channel I guess.  Ive been off the computer for bit due to Hell Week *, and there is so much new in the project since then.

*Hell Week == pulling 2 nites on-call support roster, and then having a major issue at the end of that with Amazon pulling a nation wide RDS outage for half a day.  Ouch !

andydotxyz commented:
This is looking very smooth - but I couldn't help feel that there was something about it - so I played with the numbers. See what you think about the patch I have attached - happy to chat on Slack later.

Also this bar size calculation needs more tests - and the one that exists is failing!
[inf_progress_box.diff.txt](https://github.com/fyne-io/fyne/files/2883592/inf_progress_box.diff.txt)

paulhovey commented:
> This is looking very smooth - but I couldn't help feel that there was something about it - so I played with the numbers. See what you think about the patch I have attached - happy to chat on Slack later.
> 
> Also this bar size calculation needs more tests - and the one that exists is failing!
> [inf_progress_box.diff.txt](https://github.com/fyne-io/fyne/files/2883592/inf_progress_box.diff.txt)

I like yours a lot better!  It improved the CPU usage too, or perhaps some change that upstream pulled in did :)
I fixed the test also, sorry about that!  :innocent:

andydotxyz commented on widget/infprogressbar.go:
```diff
@@ -0,0 +1,170 @@
+
+// Layout the components of the progress bar widget
+func (p *infProgressRenderer) Layout(size fyne.Size) {
+	// set height of progress bar
```
This comment seems to be erroneous

paulhovey replied:
Removed

andydotxyz commented on widget/infprogressbar.go:
```diff
@@ -0,0 +1,170 @@
+}
+
+// InfProgressBar widget creates a horizontal panel that indicates waiting indefinitely
+// An infinite progress bar loops 0% -> 100% until Stop() is called
```
probably wants "and back" after 100% with this new animation.

paulhovey replied:
Updated!

andydotxyz commented on widget/infprogressbar.go:
```diff
@@ -0,0 +1,170 @@
+// updates the infinite-style progress bar
+// can be exited by calling ProgressBar.StopInfiniteProgress()
+func (p *InfProgressBar) infiniteProgressLoop() {
+	defer p.Stop()
```
I don't think this line is needed - as Stop() must already have been called to exit the ticker range?

paulhovey replied:
`Stop()` checks to make sure the ticker is running (`p.ticker != nil`) before doing anything.  My thinking was that calling it multiple times is a lot better than not calling it at all because of the low overhead.

andydotxyz replied:
Not sure... I'd go for calling it exactly the right number of times ;)

paulhovey replied:
Looking at it with a fresh glass of beer, and you're right in that the ticker needs to stop before the defer would get hit, rendering the call useless always.  I'll remove this.

andydotxyz commented on widget/infprogressbar_test.go:
```diff
@@ -0,0 +1,54 @@
+func TestInfiniteProgressBar_Ticker(t *testing.T) {
+	bar := NewInfiniteProgressBar()
+
+	bar.Show()
```
I don't think we can rely on users calling bar.Show() - this is not the semantic of other widgets as they are visible by default...

paulhovey replied:
I removed the `Show()` call from the tests by calling `Start()` in `NewProgressBarInfinite()`.  Same effect.

andydotxyz commented on widget/infprogressbar_test.go:
```diff
@@ -0,0 +1,54 @@
+	render := Renderer(bar).(*infProgressRenderer)
+
+	// width of bar is 1/50 of width after initial creation
+	assert.Equal(t, width/50, render.bar.Size().Width)
```
I would highly recommend moving minBarWidth and maxBarWidth (with appropriate naming to avoid conflicts) to package consts so that they can be referenced in this test.
Magic numbers like this should be avoided if possible.

paulhovey replied:
This was a good suggestion, cleaned up the tests some too :)

paulhovey commented:
Man, a lot more comments and commits than I thought going into this :smile: 
I renamed the infinite progress bar (and files) to "progressbarinfinite".  It lined up nicely in the filesystem next to "progressbar", and seemed more like an extension to the progress bar rather than "infinite..."
Also cleaned up comments and made constants for the internal percentages.
